File: sdk/go/codec/codec.go
// Package codec provides payload encoding and decoding for workflow data.
//
// Built-in implementations:
//   - JSONCodec: JSON serialization
//   - BinaryCodec: raw byte passthrough
//
// Usage:
//
//	codec := codec.NewJSONCodec()
//	data, err := codec.Encode(map[string]string{"key": "value"})
//	result, err := codec.Decode(data)
package codec

import (
	"encoding/json"
	"fmt"
)

// PayloadCodec defines the interface for payload encoding/decoding.
type PayloadCodec interface {
	// Encode serializes data to bytes.
	Encode(data interface{}) ([]byte, error)

	// Decode deserializes bytes to a Go value.
	Decode(data []byte) (interface{}, error)
}

// JSONCodec encodes/decodes payloads as JSON.
type JSONCodec struct{}

// NewJSONCodec creates a new JSON codec.
func NewJSONCodec() *JSONCodec {
	return &JSONCodec{}
}

// Encode serializes data to JSON bytes.
func (c *JSONCodec) Encode(data interface{}) ([]byte, error) {
	if data == nil {
		return []byte{}, nil
	}
	if b, ok := data.([]byte); ok {
		return b, nil
	}
	return json.Marshal(data)
}

// Decode deserializes JSON bytes to a Go value.
func (c *JSONCodec) Decode(data []byte) (interface{}, error) {
	if len(data) == 0 {
		return nil, nil
	}
	var result interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("codec: JSON decode failed: %w", err)
	}
	return result, nil
}

// DecodeInto deserializes JSON bytes into a specific Go type.
func (c *JSONCodec) DecodeInto(data []byte, v interface{}) error {
	if len(data) == 0 {
		return nil
	}
	if err := json.Unmarshal(data, v); err != nil {
		return fmt.Errorf("codec: JSON decode failed: %w", err)
	}
	return nil
}

// BinaryCodec passes bytes through unchanged.
type BinaryCodec struct{}

// NewBinaryCodec creates a new binary codec.
func NewBinaryCodec() *BinaryCodec {
	return &BinaryCodec{}
}

// Encode returns the data as-is (must be []byte).
func (c *BinaryCodec) Encode(data interface{}) ([]byte, error) {
	if data == nil {
		return []byte{}, nil
	}
	b, ok := data.([]byte)
	if !ok {
		return nil, fmt.Errorf("codec: BinaryCodec expects []byte, got %T", data)
	}
	return b, nil
}

// Decode returns the data as-is.
func (c *BinaryCodec) Decode(data []byte) (interface{}, error) {
	return data, nil
}

// NullCodec encodes everything as empty bytes.
type NullCodec struct{}

// NewNullCodec creates a new null codec.
func NewNullCodec() *NullCodec {
	return &NullCodec{}
}

// Encode returns empty bytes.
func (c *NullCodec) Encode(_ interface{}) ([]byte, error) {
	return []byte{}, nil
}

// Decode returns nil.
func (c *NullCodec) Decode(_ []byte) (interface{}, error) {
	return nil, nil
}

// ChainCodec chains multiple codecs together.
// Encoding applies codecs left-to-right; decoding applies right-to-left.
type ChainCodec struct {
	codecs []PayloadCodec
}

// NewChainCodec creates a new chain codec.
func NewChainCodec(codecs ...PayloadCodec) (*ChainCodec, error) {
	if len(codecs) == 0 {
		return nil, fmt.Errorf("codec: ChainCodec requires at least one codec")
	}
	return &ChainCodec{codecs: codecs}, nil
}

// Encode applies all codecs in order.
func (c *ChainCodec) Encode(data interface{}) ([]byte, error) {
	var current interface{} = data
	for _, codec := range c.codecs {
		encoded, err := codec.Encode(current)
		if err != nil {
			return nil, err
		}
		current = encoded
	}
	if b, ok := current.([]byte); ok {
		return b, nil
	}
	return c.codecs[len(c.codecs)-1].Encode(current)
}

// Decode applies all codecs in reverse order.
func (c *ChainCodec) Decode(data []byte) (interface{}, error) {
	var current interface{} = data
	for i := len(c.codecs) - 1; i >= 0; i-- {
		decoded, err := c.codecs[i].Decode(current.([]byte))
		if err != nil {
			return nil, err
		}
		current = decoded
	}
	return current, nil
}

File: sdk/rust/src/codec.rs
//! Payload encoding/decoding codecs.
//!
//! Provides the `PayloadCodec` trait and built-in implementations:
//! - `JsonCodec` — JSON serialization via `serde_json`
//! - `BinaryCodec` — raw byte passthrough
//! - `NullCodec` — encodes everything as empty bytes
//!
//! # Example
//!
//! ```rust
//! use velocity_sdk::codec::{PayloadCodec, JsonCodec};
//!
//! let codec = JsonCodec;
//! let encoded = codec.encode(&"hello").unwrap();
//! let decoded: String = codec.decode(&encoded).unwrap();
//! assert_eq!(decoded, "hello");
//! ```

use std::fmt;

/// Trait for payload encoding and decoding.
pub trait PayloadCodec: Send + Sync {
    /// Encode a value to bytes.
    fn encode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError>;

    /// Decode bytes to a value.
    fn decode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError>;
}

/// Error type for codec operations.
#[derive(Debug, Clone)]
pub struct CodecError {
    pub message: String,
}

impl CodecError {
    pub fn new(msg: impl Into<String>) -> Self {
        Self { message: msg.into() }
    }
}

impl fmt::Display for CodecError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "CodecError: {}", self.message)
    }
}

impl std::error::Error for CodecError {}

/// JSON codec — passes through raw bytes (JSON encoding handled at application layer).
///
/// In a full implementation, this would use `serde_json` for serialization.
/// Since the engine works with raw `Vec<u8>`, this codec validates UTF-8
/// and provides a convenient wrapper.
#[derive(Debug, Clone, Copy)]
pub struct JsonCodec;

impl PayloadCodec for JsonCodec {
    fn encode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError> {
        // Validate that data is valid UTF-8 (JSON requirement)
        if !data.is_empty() {
            std::str::from_utf8(data)
                .map_err(|e| CodecError::new(format!("Invalid UTF-8 for JSON: {}", e)))?;
        }
        Ok(data.to_vec())
    }

    fn decode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError> {
        if !data.is_empty() {
            std::str::from_utf8(data)
                .map_err(|e| CodecError::new(format!("Invalid UTF-8 for JSON: {}", e)))?;
        }
        Ok(data.to_vec())
    }
}

impl JsonCodec {
    /// Encode a serializable value to JSON bytes.
    ///
    /// Requires `serde_json` to be available. This is a convenience method
    /// that serializes any `serde::Serialize` type.
    #[cfg(feature = "serde_json")]
    pub fn encode_value<T: serde::Serialize>(&self, value: &T) -> Result<Vec<u8>, CodecError> {
        serde_json::to_vec(value)
            .map_err(|e| CodecError::new(format!("JSON encode failed: {}", e)))
    }

    /// Decode JSON bytes to a deserializable value.
    #[cfg(feature = "serde_json")]
    pub fn decode_value<T: serde::de::DeserializeOwned>(&self, data: &[u8]) -> Result<T, CodecError> {
        serde_json::from_slice(data)
            .map_err(|e| CodecError::new(format!("JSON decode failed: {}", e)))
    }
}

/// Binary codec — raw byte passthrough.
#[derive(Debug, Clone, Copy)]
pub struct BinaryCodec;

impl PayloadCodec for BinaryCodec {
    fn encode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError> {
        Ok(data.to_vec())
    }

    fn decode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError> {
        Ok(data.to_vec())
    }
}

/// Null codec — encodes everything as empty bytes.
///
/// Useful for workflows that take no input and return no output.
#[derive(Debug, Clone, Copy)]
pub struct NullCodec;

impl PayloadCodec for NullCodec {
    fn encode(&self, _data: &[u8]) -> Result<Vec<u8>, CodecError> {
        Ok(Vec::new())
    }

    fn decode(&self, _data: &[u8]) -> Result<Vec<u8>, CodecError> {
        Ok(Vec::new())
    }
}

/// Chain multiple codecs together.
///
/// Encoding applies codecs left-to-right; decoding applies right-to-left.
pub struct CodecChain {
    codecs: Vec<Box<dyn PayloadCodec>>,
}

impl CodecChain {
    /// Create a new codec chain.
    pub fn new(codecs: Vec<Box<dyn PayloadCodec>>) -> Result<Self, CodecError> {
        if codecs.is_empty() {
            return Err(CodecError::new("CodecChain requires at least one codec"));
        }
        Ok(Self { codecs })
    }
}

impl PayloadCodec for CodecChain {
    fn encode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError> {
        let mut current = data.to_vec();
        for codec in &self.codecs {
            current = codec.encode(&current)?;
        }
        Ok(current)
    }

    fn decode(&self, data: &[u8]) -> Result<Vec<u8>, CodecError> {
        let mut current = data.to_vec();
        for codec in self.codecs.iter().rev() {
            current = codec.decode(&current)?;
        }
        Ok(current)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_json_codec_roundtrip() {
        let codec = JsonCodec;
        let data = br#"{"key": "value"}"#;
        let encoded = codec.encode(data).unwrap();
        let decoded = codec.decode(&encoded).unwrap();
        assert_eq!(decoded, data);
    }

    #[test]
    fn test_json_codec_invalid_utf8() {
        let codec = JsonCodec;
        let data = &[0xFF, 0xFE];
        assert!(codec.encode(data).is_err());
    }

    #[test]
    fn test_binary_codec_roundtrip() {
        let codec = BinaryCodec;
        let data = vec![0u8, 1, 2, 3, 255];
        let encoded = codec.encode(&data).unwrap();
        let decoded = codec.decode(&encoded).unwrap();
        assert_eq!(decoded, data);
    }

    #[test]
    fn test_null_codec() {
        let codec = NullCodec;
        let data = b"anything";
        let encoded = codec.encode(data).unwrap();
        assert!(encoded.is_empty());
        let decoded = codec.decode(data).unwrap();
        assert!(decoded.is_empty());
    }

    #[test]
    fn test_codec_chain() {
        let chain = CodecChain::new(vec![
            Box::new(JsonCodec),
            Box::new(BinaryCodec),
        ]).unwrap();

        let data = br#"{"test": true}"#;
        let encoded = chain.encode(data).unwrap();
        let decoded = chain.decode(&encoded).unwrap();
        assert_eq!(decoded, data);
    }

    #[test]
    fn test_empty_chain_error() {
        assert!(CodecChain::new(vec![]).is_err());
    }
}
